🚀 4단계 - 페이먼츠(카드 수정)

no1msh opened the pull request:
## 4단계 - 페이먼츠(카드 수정) 기능 요구 사항
- [x] 카드 수정 기능을 구현한다.
  - [x] 카드 수정 화면에서 변경사항이 발생하지 않으면 수정이 불가하다.
  - [x] 카드가 수정되면 카드 목록 화면에 변경사항이 반영된다.

---

안녕하세요 영직님 4단계 미션 해왔습니다!
쉬울줄 알았는데 생각보다 생각할 거리가 많네요!
제가 궁금한 사항은 다음과 같습니다.

---
### UI State가 sealed class or interface로 작성되었을 때 타입 캐스팅에 따른 불편함을 어떻게 해소하시나요?

- MutableStateFlow의 경우 가변프로퍼티로 처리되어 컴파일러가 스마트 캐스팅을 지원하지 않습니다.
- 따라서 저는 아래와 같이 지역변수에 담아서 분기처리를 해주었는데요.

```kotlin
when (val state: UpdateCardUiState = _uiState.value) {
    is UpdateCardUiState.AddCardUiState -> _uiState.update {
        state.copy(cardState = state.cardState.copy(password = newPassword))
    }

    is UpdateCardUiState.EditCardUiState -> _uiState.update {
        state.copy(cardState = state.cardState.copy(password = newPassword))
    }
}
```

- 이렇게 할 경우 컴파일러가 우려하던 MutableStateFlow의 갑작스런 값의 변화에 대응하지 못할 것 같은데, 혹시 더 좋은 방법이 있을까요?

no1msh committed:
docs: 4단계 미션 요구사항 정의

no1msh committed:
feat: 결제 카드 레포지토리는 카드를 수정할 수 있다

no1msh committed:
refactor: 카드 객체 parcelable 객체로 수정

no1msh committed:
feat: Parcelable 유틸함수 구현

no1msh committed:
refactor: CardInformationInputFields 컴포넌트 분리

no1msh committed:
refactor: UpdateCardTopBar에 제목과 버튼 활성화를 파라미터로 받게 변경

no1msh committed:
refactor: VisualTransformation 패키지 이동

no1msh committed:
feat: intent를 통한 카드 객체 전달

no1msh committed:
refactor: 기존 NewCard~ 를 UpdateCard로 변경

no1msh committed:
feat: UpdateCardUiState 구현

no1msh committed:
refactor: 카드 추가 화면 컴포넌트 분리

no1msh committed:
feat: 카드 수정 화면 컴포넌트 구현

no1msh committed:
feat: 카드 클릭 이벤트 구현

no1msh committed:
feat: 카드 수정 기능 구현

no1msh committed:
test: UpdateCardScreen은 UiState에 따라 다른 Screen을 표시한다.

no1msh committed:
test: AddCardScreen은 카드 정보 입력 여부에 따라 카드 추가 버튼이 활성화된다

no1msh committed:
test: EditCardScreen은 수정하려는 카드의 변경사항이 없다면 카드 수정을 완료할 수 없다

no1msh committed:
docs: 기능목록 구현완료현황 최신화

no1msh committed:
refactor: 탑바의 enabled의 대상이 완료 버튼임을 명시

no1msh committed:
fix: 테스트 함수명 오류 개선

no1msh committed:
refactor: 테스트코드의 불필요 구문 제거

no1msh committed:
refactor: 인텐트를 가져오는 함수명을 좀더 의도를 알기쉽게 변경

no1msh committed:
refactor: 카드와 카드상태 객체간 매퍼 구현

no1msh committed:
refactor: 카드사 매퍼 패키지 이동

no1msh committed:
refactor: 카드 상태 객체 분리

no1msh committed:
refactor: 카드 도메인 객체를 전달하는 것이 아닌 카드 상태 객체를 전달하도록 변경

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/components/UpdateCardTopBar.kt:
```diff
@@ -35,7 +37,7 @@ fun NewCardTopBar(
         },
         actions = {
-            IconButton(onClick = onSaveClick) {
+            IconButton(onClick = onSaveClick, enabled = doneButtonEnabled) {
```
마이너한 내용인데 save와 done이 서로 다른 이유가 있나요~?
이렇게 나열하면 사용자가 더 알기 쉽지 않을 까 해서요
```kotlin
saveEnabled: Boolean,
onSaveClick: () -> Unit,
```

no1msh replied:
불필요하게 다른 부분인 것 같아 수정하였습니다!
[커밋](https://github.com/next-step/android-payments/pull/107/commits/535fce35e39c503b689177c49931cc4e52216b52)

laco-dev commented on app/src/androidTest/java/nextstep/payments/AddCardScreenTest.kt:
```diff
@@ -0,0 +1,82 @@
+package nextstep.payments
```
컴포저블 테스트 패키지도 단위 테스트와 마찬가지로 구현 코드와 동일하게 맞춰주세요!

no1msh replied:
수정하였습니다!
[커밋](https://github.com/next-step/android-payments/pull/107/commits/c954266fe06daf04ffda9f3755b6a8026acfbde8)

laco-dev commented on app/src/androidTest/java/nextstep/payments/EditCardScreenTest.kt:
```diff
@@ -0,0 +1,95 @@
+                    password = "1234",
+                    selectedCardCompany = CardCompanyState.HANA
+                )
+                EditCardScreen(
+                    uiState = UpdateCardUiState.EditCardUiState(
+                        cardForEdit = card,
+                        cardState = card,
+                        cardUpdated = false
+                    ),
+                    onCardCompanyClick = {},
+                    onCardNumberChange = {},
+                    onExpiredDateChange = {},
+                    onOwnerNameChange = {},
+                    onPasswordChange = {},
+                    onBackClick = {},
+                    onSaveClick = {},
+                )
```
단위 테스트와 비슷하게 테스트를 위한 가독성 개선 차원에서 이런 식으로 스펙을 사용할 수 있겠네요!
나머지 람다 함수는 관심사가 아니니까요!
```kotlin
private fun EditCardScreen(
   ...
   onCardCompanyClick = {},
)
```

no1msh replied:
모든 컴포저블 함수의 람다함수 파라미터를 기본값을 {}로 지정할까요?

만일 그렇다하시면 필수로 넘겨줘야할 람다 파라미터들도 사용하는 쪽에서 람다를 넘겨주지 않아도 컴파일 에러가 나지 않아 휴먼에러를 발생할 염려가 되는데 괜찮을까요?

제가 영직님 코멘트의 의도를 잘 못파악한 걸 수도 있으니 편하게 다시 말씀 부탁드리겠습니다!

laco-dev replied:
아~ 제 의도는 테스트코드에서만 사용하기 위한 함수를 추가로 만드는 것으로 생각했습니다.
구현 코드는 유지하고요!

no1msh replied:
아아 이해하였습니다. 그렇게한다면 테스트 코드 가독성을 훨신 높일 수 있겠군요!! 감사합니다.

laco-dev commented on app/src/androidTest/java/nextstep/payments/EditCardScreenTest.kt:
```diff
@@ -0,0 +1,95 @@
+                val password = "1234"
+
+                EditCardScreen(
+                    uiState = UpdateCardUiState.EditCardUiState(
+                        cardForEdit = CardState(
+                            cardNumber = cardNumbers,
+                            expiredDate = expiredDate,
+                            ownerName = ownerName,
+                            password = password,
+                            selectedCardCompany = CardCompanyState.HANA
+                        ), cardState = CardState(
+                            cardNumber = cardNumbers,
+                            expiredDate = expiredDate,
+                            ownerName = ownerName,
+                            password = password,
+                            selectedCardCompany = CardCompanyState.KAKAO
+                        ),
+                        cardUpdated = false
+                    ),
```
한편으로는 EditCardScreen이 previous 상태를 알 필요가 있을까요?
UI 입장에서는 수정이 가능한지 불가능한지 여부만 알고싶지 이전 카드의 상태가 필요한게 아니니까요.

그렇다면 이 필드값은 어디서 만들어서 넣어주는게 좋을까요?

no1msh replied:
많이 고민이 되었던 사항이지만, UI State가 표시할 것도 아닌데 수정을 할 카드에 대한 상태를 전부 알고 있는 것이 어색함이 공감됩니다!

카드의 수정사항이 있는지 체크하는 로직의 책임을 누가가져야 할까 고민을 많이 해보았지만,
뷰모델이 수정을 할 카드의 상태를 가지고 있고, 현재 카드 상태 또한 가지고 있기 떄문에 둘의 비교를 위해서, 일단은 뷰모델의 메소드로 `canEdit()`함수를 구현하였습니다.

그치만... 뷰모델에 값을 비교하는 로직이 있는 것이 과연 맞는가 싶은 마음도 듭니다...
더 좋은 개선 방향이 있을까요?

no1msh replied:
[커밋](https://github.com/next-step/android-payments/pull/107/commits/12c2033cd0d9e0e2c9ac1a581331f04a68a130c2)

laco-dev replied:
왜 맞지 않다고 생각하시는지 궁금하네요.
상태를 관리하는 역할로 본다면 적절한 위치가 아닐까 생각하는데요?

no1msh replied:
가급적이면 뷰모델에게 부가적인 책임을 주기 싫었던 마음에 언급하였습니다!

다시 고려한 결과 뷰모델이 제일 어울리는 것 같습니다!

laco-dev commented on app/src/main/java/nextstep/payments/components/card/BaseCard.kt:
```diff
@@ -23,23 +22,50 @@ fun BaseCard(
-@Preview(showBackground = true)
 @Composable
-private fun BaseCardPreview() {
+fun BaseCard(
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier,
+    color: Color = Black100,
+    content: @Composable ColumnScope.() -> Unit,
```
위에 있는 BaseCard와 스펙이 동일한데 클릭 기능만 추가된 것으로 보여요.

컴포즈에서 이런 기능 차이를 여러개의 함수로 나누는 것도 방법이지만 단순히 Modifier만 변경되는 정도라면 then을 이용하고는 합니다.
```kotlin
fun BaseCard(
  onClick: (() -> Unit)? = null,
  ...
) {
  modifier = modifier.then(
    if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier
  )
   ...
}
```

no1msh replied:
`BaseCard` 컴포넌트에 클릭이벤트를 추가하면서 `Surface` Component를 활용하게 되었습니다.
`Modifier`를 통해 `onClick`을 지정하게 되면 Radius가 지정되어있기 때문에 리플 효과의 영역이 올바르게 표시되지 않는 현상이 있어 `clip`도 같이 지정을 해줘야하기 때문입니다.

더불어 `Surface`의 내부 코드를 확인해보니까 여기서도 `onClick()`을 기준으로 두 컴포넌트로 나눠져 구현이 되어있더라구요!

**이대로 유지해도 괜찮을까요?**

`then()`은 잘 안써봐서 이번기회에 알아봤는데, 상당히 유용하군요!
- `Modifier`를 조건에 따라 지정할 때
- `Modifier`의 순서를 조정할 때

laco-dev replied:
네 Surface는 onClick이 다른 함수로 되어있어서 번거롭겠네요.
then() 함수는 조건에 따라 Modifier를 결정하기 위해서 사용합니다 :)

필요할 때 적절히 사용해보시는것으로 충분합니다.

laco-dev commented on app/src/main/java/nextstep/payments/domain/PaymentCardsRepository.kt:
```diff
@@ -4,4 +4,6 @@ interface PaymentCardsRepository {
     fun getCards(): List<Card>
 
     fun addCard(card: Card)
+
+    fun updateCard(oldCard: Card, newCard: Card)
```
🤔  아키텍쳐와 관련된 내용을 다루지는 않지만, 궁금해서 여쭤봅니다.

카드를 업데이트 하는데 이전 카드를 같이 줄 필요가 있나요?
현재 넘겨주는 카드에 아이디와 같은 식별자가 있지 않은가 해서요!

no1msh replied:
카드 객체에 Id를 할당하지 않았고, 그렇기 떄문에 카드끼리의 프로퍼티 비교를 통해 수정할 카드를 레포지토리로부터 찾을 수 있어 카드 두개를 인자로 받게끔 구현 하였었습니다!

수정사항으론 카드 객체에 ID를 부여하였고, `fun updateCard(newCard: Card)` 시그니처로 변경하였습니다.
이제 인자로 들어온 Card 객체의 ID를 기준으로 카드를 찾고 새로 들어온 카드로 변경합니다.

no1msh replied:
[커밋](https://github.com/next-step/android-payments/pull/107/commits/897c7c06c645d99a0d5e31a93a55eddf45490e17)

no1msh replied:
한편으로는 Card 도메인 모델에 Id가 부여되고 Card <-> CardUiState 로 변환되면서 CardUiState에도 Id가 생기게 되었는데,
UI의 상태와는 거리가 멀다 느껴지는데 괜찮을까요?
LazyColumn의 Key로 사용이 될 수 있으니 괜찮아 보이기도 하고 고민이 됩니다.

laco-dev replied:
CardUiState가 Card를 가지고 있는건 자연스러운 방향이 아닐까요?
왜 UI 상태가 도메인을 알면 안되는 것인지요?

디자인에서 수정을 하면 매번 UI 상태를 바꿔야하는 걸까요?
저라면 도메인 모델을 그대로 넘겨주어 디자인 쪽에서 자유롭게 바꿀 수 있는게 더 좋다고 생각해서요

no1msh replied:
id가 UI의 상태가 될 수 있냐? 라는 질문을 스스로 해보았을 때 뭔가 어색한 느낌을 받았거든요!

> CardUiState가 Card를 가지고 있는건 자연스러운 방향

이라는 말을 들으니 납득이 되었습니다.

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/list/CardListUiState.kt:
```diff
@@ -1,9 +1,9 @@
     data object Empty : CardListUiState
-    data class One(val card: Card) : CardListUiState
-    data class Many(val cards: List<Card>): CardListUiState
+    data class One(val card: CardState) : CardListUiState
+    data class Many(val cards: List<CardState>) : CardListUiState
```
저는 Card가 도메인 모델이라는 역할로 생각했는데요. 그렇지 않군요 🤔 
CardList는 UiState인데 Card는 State로 지은 이유가 있을까요?

no1msh replied:
UiState와 State라는 이름을 혼용해서 사용해서 혼란을 드린 것 같습니다.  🥲

`Card`가 도메인 모델, `CardUiState`가 카드에 해당하는 UI State 입니다!

전체적으로 코드 확인후 파라미터명, 타입명 맞게 변경하였습니다!

[커밋](https://github.com/next-step/android-payments/pull/107/commits/9962d318ed090a270918757bbf11d4ff2708d351)

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/EditCardScreen.kt:
```diff
@@ -0,0 +1,123 @@
+    onSaveClick: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    var showCardCompanyBottomSheet by remember { mutableStateOf(true) }
```
수정을 위한 화면인데 카드사 선택이 최초 1회 강제로 되는게 조금 불편하지 않을까요?

no1msh replied:
맞습니다!
카드를 다시 클릭하면 바텀시트가 올라와 다시 한번 카드사를 고를 수 있게 조치하였습니다.

no1msh replied:
[커밋](https://github.com/next-step/android-payments/pull/107/commits/53fc9de46042e10c3045809dbbf83a5f3a9dc5b8)

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/EditCardScreen.kt:
```diff
@@ -0,0 +1,123 @@
+        modifier = modifier,
+        containerColor = Color.White,
+    ) { innerPadding ->
+        if (showCardCompanyBottomSheet) {
+            CardCompanyBottomSheetDialog(
+                onDismissRequest = { showCardCompanyBottomSheet = false },
+                onCardCompanyClick = onCardCompanyClick,
+            )
+        }
```
크게 중요한 부분은 아닙니다.

바텀시트의 경우 Popup에서 동작하기 때문에 어디에 위치해도 동일하게 작동하는 효과가 있는데요.
Scaffold의 content에 넣는 것과 옆에 넣는 것 어떻게 생각하시나요?

no1msh replied:
넵 어디에 위치해도 동일하니 컴포넌트의 내용에 해당하는 content에 있을 필요가 없다고 판단되어 수정하였습니다!

no1msh replied:
[EditCardScreen쪽 커밋](https://github.com/next-step/android-payments/pull/107/commits/5ae24b4fb9c56b0b79f2b7dc1b3ca0eba6e343d2)

[AddCardScreen쪽 커밋](https://github.com/next-step/android-payments/pull/107/commits/544dee38a8f7384475c18b9dbdcc9f2e837185ec)

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/UpdateCardActivity.kt:
```diff
@@ -0,0 +1,44 @@
+                UpdateCardScreen(
+                    cardState = cardState,
+                    onBackClick = { finish() },
+                    navigateToCardList = {
+                        setResult(RESULT_OK)
+                        finish()
+                    }
```
카드 편집이 완료되었다는 동작은 
`ViewModel -> Activity`로 전파하는 것과 `ViewModel -> Screen -> Activity`로 전파하는 차이가 있는데요.
어떤 방식이 좋다고 보시나요?

no1msh replied:
카드 편집이 완료되었다는 사실을 Screen 단위에서 이용하는 것이 없이 단순 넘겨주기만 하기 때문에 
`ViewModel -> Activity`로 전파하는 형태로 간추렸습니다.

no1msh replied:
[커밋](https://github.com/next-step/android-payments/pull/107/commits/522f036cafbe007e80c38be935dc060141f62dcb)

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/UpdateCardScreen.kt:
```diff
@@ -0,0 +1,82 @@
+        if (uiState.cardUpdated) navigateToCardList()
+    }
+
+    LaunchedEffect(Unit) {
+        if (cardState != null) {
+            viewModel.setEditCardMode(cardState)
+        }
+    }
```
고민해볼만한 주제인데요.
ViewModel의 init 함수에서 이 동작을 스스로 하는 것과, LaunchedEffect에서 호출하는 것의 차이점을 아시나요?

no1msh replied:
ViewModel의 init에서 한다면 예를들어 Route단 컴포넌트에서 viewModel 기본인자로 `viewModel()` or `hiltViewModel()`이 아니라 액티비티 같은 곳에서 ViewModel 인스턴스로 만들고 컴포넌트에 넘겨주는 상황이라면
Screen 함수에 진입할 때 초기화 되지 않는 현상이 있을 것 같습니다.

반대로 LauchedEffect에서 호출한다면 해당 이펙트 함수가 있는 컴포저블이 렌더링 될 떄 호출 될 것 같습니다.

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/UpdateCardUiState.kt:
```diff
@@ -0,0 +1,24 @@
+        }
+    }
+
+    data class AddCardUiState(
+        override val cardState: CardState = CardState(),
+        override val cardUpdated: Boolean = false,
+    ) : UpdateCardUiState
```
이미 UiState 타입이라서 상속하는 곳에서도 UiState를 굳이 붙이지 않아도 되지 않을까? 라는 생각이 듭니다!

no1msh replied:
어차피 임포트를 따로 하지 않는 이상 `UpdateCardUiState.Add` 이런식으로 작성하기 떄문에 말씀 주신 것처럼 변경해보았습니다!

no1msh replied:
[커밋](https://github.com/next-step/android-payments/pull/107/commits/e534b712b6a57b80d51d636538718f34c05fb8ce)

laco-dev commented on app/src/main/java/nextstep/payments/screens/card/update/UpdateCardViewModel.kt:
```diff
@@ -0,0 +1,132 @@
+        }
+    }
+
+    fun setCardNumber(newCardNumber: String) {
+        if (newCardNumber.length > MAX_CARD_NUMBER_LENGTH) return
+
+        when (val state: UpdateCardUiState = _uiState.value) {
+            is UpdateCardUiState.AddCardUiState -> _uiState.update {
+                state.copy(cardState = state.cardState.copy(cardNumber = newCardNumber))
+            }
+
+            is UpdateCardUiState.EditCardUiState -> _uiState.update {
+                state.copy(cardState = state.cardState.copy(cardNumber = newCardNumber))
+            }
+        }
+    }
```
> UI State가 sealed class or interface로 작성되었을 때 타입 캐스팅에 따른 불편함을 어떻게 해소하시나요?

트레이드 오프의 영역이라고 생각해요.
말씀하신 것처럼 캐스팅에 따른 불편이 분명히 존재합니다.
명확하게 타입이 나뉘어야 하는지 그렇지 않고 data class의 장점을 이용해 한번에 표현할 것인지 상황마다 다르기에 어려운 내용 같습니다.

다만 저는 개인적으로 화면에 대한 Root 성격을 담당하는 UiState는 항상 data class로 작성을 합니다.

no1msh replied:
으음 확실히 은탄환은 없군요 ㅎㅎ

그러면 만약 카드 추가 화면과 카드 수정 화면을 각각의 다른 화면과 기능이라고 판단하고,

아에 다른 뷰모델에서 각각의 UiState를 data class로 사용하는 편이 나았을까요?

비슷해보이는 두 화면을 둘로 나누면서 생기는 DRY 원칙 위배는 컴포넌트 재사용으로 어느정도 해소가 가능할 것 같아서요!

laco-dev replied:
너무 원칙적인 부분을 챙기시기 보다는 실무나 다른 프로젝트에 적용을 해 보시면서 자신만의 방법을 찾는게 좋아보여요.
말씀하신 것처럼 방법마다 장단점이 명확히 있다보니 원칙을 매번 챙기면 팀원들을 설득하기 어렵습니다.

no1msh replied:
좋은 말씀 너무 감사드립니다 ㅎㅎ

laco-dev reviewed (changes requested):
4단계 진행 고생 많으셨습니다.

이것저것 고민하시고 만드신 흔적이 보이네요.
질문주신 내용 포함해서 몇가지 의견 남겨드렸습니다! 🙏🏻

no1msh committed:
refactor: 완료 버튼을 저장 버튼으로 이름 변경

no1msh committed:
refactor: 테스트 코드들의 패키지를 구현코드와 동일하게 변경

no1msh committed:
refactor: 카드 객체, UI State에 ID값 부여

- ID 할당은 Repository 구현체에서 담당
- CardListActivity -> UpdateCardActivity로 향하는 Intent는 CardUiState 객체가 아닌 Card의 ID 값으로 변경
- UpdateCardActivity에서 UpdateCardViewModel로 intent로 받은 Card ID 전달
- Parcelable 의존성 제거

no1msh committed:
refactor: UI State에서 수정을 할 카드 상태를 제거

no1msh committed:
refactor: State 이름을 UiState로 변경

no1msh committed:
refactor: 바텀시트 다이얼로그를 Scaffold content 스코프영역에서 옆으로 이동

no1msh committed:
refactor: 카드의 편집 완료 유무를 UiState에서 제거

- ViewModel -> Activity 방향으로 전파

no1msh committed:
refactor: 중복되는 접미사 UiState 제거

no1msh committed:
refactor: 사용하지 않는 코드 제거

no1msh committed:
refactor: 바텀 시트 다이얼로그를 Scaffold의 content 스코프에서 옆으로 이동

no1msh committed:
feat: 카드를 추가/수정 할 때 카드 컴포넌트를 클릭하면 카드사를 재선택 할 수 있다.

laco-dev reviewed (approved):
피드백 반영 고생하셨습니다.

코드를 추가로 수정하지는 않아도 될 것 같아서 미션은 마무리하고, 질문에만 답변을 조금 남겨두었습니다.

남은 기간동안 좋은 경험 더 받아가시면 좋겠습니다~
